### DISCUSSION TITLE
mini-css-extract-plugin Conflicting order warning in gatsby v3

### DISCUSSION THREAD

**Discussion by shameer-rahman** [gatsbyjs/gatsby#30169] [Help]
I'm using the following webpack config to silence the css order warning which was working fine until i upgraded to gatsby v3.

```
exports.onCreateWebpackConfig = ({ stage, actions, getConfig, loaders }) => {
	if (stage === 'build-javascript') {
	  const config = getConfig()
	  const miniCssExtractPlugin = config.plugins.find(
		plugin => plugin.constructor.name === 'MiniCssExtractPlugin'
	  )
	  if (miniCssExtractPlugin) {
		miniCssExtractPlugin.options.ignoreOrder = true
	  }
	  actions.replaceWebpackConfig(config)
	}
  }
```

I'm getting the following errors

```
warn chunk commons [mini-css-extract-plugin]
Conflicting order. Following module has been added:
 * css ./node_modules/css-loader/dist/cjs.js??ruleSet[1].rules[9].oneOf[0].use[1]!./node_modules/postcss-loader/dist/cjs.js??ruleSet[1].rul
es[9].oneOf[0].use[2]!./node_modules/sass-loader/dist/cjs.js??ruleSet[1].rules[9].oneOf[0].use[3]!./src/components/ContentfulModules/Module
Image.module.scss
despite it was not able to fulfill desired ordering with these modules:
 * css ./node_modules/css-loader/dist/cjs.js??ruleSet[1].rules[9].oneOf[0].use[1]!./node_modules/postcss-loader/dist/cjs.js??ruleSet[1].rul
es[9].oneOf[0].use[2]!./node_modules/sass-loader/dist/cjs.js??ruleSet[1].rules[9].oneOf[0].use[3]!./src/components/ModuleCategoriesLi
st.module.scss
   - couldn't fulfill desired order of chunk group(s) component---src-templates-page-about-tsx
   - while fulfilling desired order of chunk group(s) component---src-templates-page-contact-tsx
```

however the above code doesn't seem to work anymore, is there anyway to hide the warnings?

**Comment by chris-erickson** [comment]
I [created an issue](https://github.com/mediacurrent/gatsby-plugin-silence-css-order-warning/issues/1) for this, which is unresolved.

**Comment by kamranayub** [comment]
FWIW I did not have any overrides/changes to webpack config in v2 and now I get these in V3, so I am not sure what that means (whether they were there before and silenced automatically or whether something has changed in the way things work). It doesn't _appear_ to be causing any runtime issues.

**Reply by ediblecode** [reply]
I'm finding exactly the same - v2 was fine (no warnings, CSS as expected) and v3 I have all the warnings and likewise - the site itself seems fine.

**Reply by dungle-scrubs** [reply]
I'm getting the same warnings.  I'm using Material UI for almost everything and these warnings show up specifically for chunks on pages where I import `.scss` files.

**Comment by juanbiberretta** [comment]
Interesting, I'm running on Gatsby 3.3 with this snippet and I was able to ignore them: 

```js
  if (stage === 'build-javascript' || stage === 'develop') {
    const config = getConfig();

    const miniCss = config.plugins.find(
      (plugin) => plugin.constructor.name === 'MiniCssExtractPlugin'
    );

    if (miniCss) {
      miniCss.options.ignoreOrder = true;
    }

    actions.replaceWebpackConfig(config);

    actions.setWebpackConfig({
      plugins: [plugins.provide({ process: 'process/browser' })],
    });
  }
```

Doesn't look much different to yours with the exception of doing it on `stage === 'develop'` - are you seeing this on `gatsby build` _and_ `gatsby develop` ?

**Comment by wardpeet** [comment]
:wave:

You probably want to remove the plugin and re-create it. This way you're certain that everything is set correctly. (this snippet could have syntax errors)

```js
exports.onCreateWebpackConfig = ({ stage, actions, getConfig, loaders, plugins }) => {
	  const config = getConfig()
	  const miniCssExtractPluginIndex = config.plugins.findIndex(
		plugin => plugin.constructor.name === 'MiniCssExtractPlugin'
	  )

	  if (miniCssExtractPluginIndex > -1) {
        // remove miniCssExtractPlugin from plugins list
        config.plugins.splice(miniCssExtractPluginIndex, 1);

        // re-add mini-css-extract-plugin
        if (stage === 'build-javascript') {
          config.plugins.push(plugins.extractText({
            filename: `[name].[contenthash].css`,
            chunkFilename: `[name].[contenthash].css`,
            ignoreOrder: true,
          }))
        } else {
          config.plugins.push(plugins.extractText({
            filename: `[name].css`,
            chunkFilename: `[id].css`,
            ignoreOrder: true,
          }))
        }
	  }
	  actions.replaceWebpackConfig(config)
	}
  }
```

**Reply by TeoTN** [reply]
Why is this required? I mean, it worked perfectly in v2

**Reply by wardpeet** [reply]
You don't have to, with this approach you're 100% sure that it's a new plugin instead of changing private variables of the existing one.

**Comment by hdorgeval** [comment]
Hi, I just got this warning in Gatsby 3.7 while I started refactoring my react components by moving them to a different folder hierarchy.
Though all export/import declarations were correctly updated after each move, I got this warning at each refactoring step.
Then those warnings disappeared, as soon as my refactor was finished and all my export barrels had been revamped.

So I imagine that if you got this warning, maybe it is because there is some overlapping in the barrel files from which you import your components, overlapping that could not be detected by the typescript compiler, but only by the css plugin.

**Comment by renet** [comment]
Ok, this is a year ago now, but after upgrading to Gatsby 4 lately (legacy project), I started getting these warnings, as well.

My questions on this topic:

1. Except from silencing/ignoring these issues, can anyone shed some light on the actual reasons for these warnings?
2. Does anyone know why this was not an issue with Gatsby v2?
3. Is @wardpeet's solution the best possible solutions for this, or is there a cleaner way to solve this?
4. I think the following message I get during the `gatsby build` may be related. Can anyone help me with this:

```
<w> [webpack.cache.PackFileCacheStrategy] Skipped not serializable cache item 'mini-css-extract-plugin /Users/M305642/dev/ux-toolkit/node_modules/gatsby/node_modules/css-loader/dist/cjs.js??ruleSet[1].rules[11].oneOf[1].use[1]!/Users/M305642/dev/ux-toolkit/node_modules/gatsby/node_modules/postcss-loader/dist/cjs.js??ruleSet[1].rules[11].oneOf[1].use[2]!/Users/M305642/dev/ux-toolkit/node_modules/sass-loader/dist/cjs.js??ruleSet[1].rules[11].oneOf[1].use[3]!/Users/M305642/dev/ux-toolkit/src/components/searchResults/searchResults.sass|0|Compilation/modules|/Users/M305642/dev/ux-toolkit/node_modules/gatsby/node_modules/css-loader/dist/cjs.js??ruleSet[1].rules[11].oneOf[1].use[1]!/Users/M305642/dev/ux-toolkit/node_modules/gatsby/node_modules/postcss-loader/dist/cjs.js??ruleSet[1].rules[11].oneOf[1].use[2]!/Users/M305642/dev/ux-toolkit/node_modules/sass-loader/dist/cjs.js??ruleSet[1].rules[11].oneOf[1].use[3]!/Users/M305642/dev/ux-toolkit/src/components/searchResults/searchResults.sass': No serializer registered for Warning
<w> while serializing webpack/lib/cache/PackFileCacheStrategy.PackContentItems -> webpack
```
